cmd/plugin/cmd/upgrade_plan_test.go

```diff
@@ -0,0 +1,179 @@
+				targetNamespace: "capi-operator-system",
+			},
+		},
+		{
+			name:      "custom infra provider",
+			customURL: "https://github.com/kubernetes-sigs/cluster-api/releases/latest/core-components.yaml",
+			wantedUpgradePlan: upgradePlan{
+				Contract: "v1beta1",
+				Providers: []upgradeItem{
+					{
+						Name:           "docker",
+						Namespace:      "capi-system",
+						Type:           "infrastructure",
+						CurrentVersion: "v1.6.0",
+						Source:         "https://github.com/kubernetes-sigs/cluster-api/releases/latest/core-components.yaml",
+						SourceType:     providerSourceTypeCustomURL,
+					},
+				},
+			},
+			wantedProviders: []genericprovider.GenericProvider{
+				generateGenericProvider(clusterctlv1.InfrastructureProviderType, "docker", "capi-system", "v1.6.0", "", ""),
+			},
+			wantErr: false,
+			opts: &initOptions{
+				infrastructureProviders: []string{"docker:capi-system:v1.6.0"},
+				targetNamespace:         "capi-operator-system",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			resources := []ctrlclient.Object{}
+
+			for _, provider := range tt.wantedProviders {
+				resources = append(resources, provider)
+			}
+
+			err := initProviders(ctx, env, tt.opts)
+			if tt.wantErr {
+				g.Expect(err).To(HaveOccurred())
+
+				return
+			} else {
+				g.Expect(err).NotTo(HaveOccurred())
+			}
+
+			for _, genericProvider := range tt.wantedProviders {
+				g.Eventually(func() (bool, error) {
+					provider, err := getGenericProvider(ctx, env, string(util.ClusterctlProviderType(genericProvider)), genericProvider.GetName(), genericProvider.GetNamespace())
+					if err != nil {
+						return false, err
+					}
+
+					if provider.GetSpec().Version != genericProvider.GetSpec().Version {
+						return false, nil
+					}
+
+					return true, nil
+				}, waitShort).Should(BeTrue())
+			}
+
+			// Init doesn't support custom URLs yet, so we have to update providers here
+			if tt.customURL != "" {
+				for _, genericProvider := range tt.wantedProviders {
+					provider, err := getGenericProvider(ctx, env, string(util.ClusterctlProviderType(genericProvider)), genericProvider.GetName(), genericProvider.GetNamespace())
+					g.Expect(err).NotTo(HaveOccurred())
+
+					spec := provider.GetSpec()
+					spec.FetchConfig = &operatorv1.FetchConfiguration{
+						URL: tt.customURL,
+					}
+
+					provider.SetSpec(spec)
+
+					g.Expect(env.Update(ctx, provider)).NotTo(HaveOccurred())
+
+					g.Eventually(func() (bool, error) {
+						provider, err := getGenericProvider(ctx, env, string(util.ClusterctlProviderType(genericProvider)), genericProvider.GetName(), genericProvider.GetNamespace())
+						if err != nil {
+							return false, err
+						}
+
+						if provider.GetSpec().FetchConfig == nil || provider.GetSpec().FetchConfig.URL != tt.customURL {
+							return false, nil
+						}
+
+						return true, nil
+					}, waitShort).Should(BeTrue())
```

Comment:
This format makes things a bit more tidy IMO.
```suggestion
					g.Eventually(func(g g.Gomega) error {
						provider, err := getGenericProvider(ctx, env, string(util.ClusterctlProviderType(genericProvider)), genericProvider.GetName(), genericProvider.GetNamespace())
						g.Expect(err).ToNot(HaveOccurred())
						g.Expect(provider.GetSpec().FetchConfig != nil && provider.GetSpec().FetchConfig.URL == tt.customURL).To(BeTrue())
					}, waitShort).Should(Succeed())
```